server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -150,32 +153,39 @@ public void run() {
-                try {
-                    if (Files.exists(path)) {
-                        Path tempDataPath = path.resolve(TEMP_FILE_NAME);
-                        Files.deleteIfExists(tempDataPath);
-                        try (OutputStream os = Files.newOutputStream(tempDataPath, StandardOpenOption.CREATE_NEW)) {
-                            os.write(byteToWrite);
-                            IOUtils.fsync(tempDataPath, false);
+            boolean lockAssertionFailed = false;
+            try {
+                for (Path path : nodeEnv.nodeDataPaths()) {
+                    long executionStartTime = currentTimeMillisSupplier.getAsLong();
+                    try {
+                        if (Files.exists(path)) {
+                            Path tempDataPath = path.resolve(TEMP_FILE_NAME);
+                            Files.deleteIfExists(tempDataPath);
+                            try (OutputStream os = Files.newOutputStream(tempDataPath, StandardOpenOption.CREATE_NEW)) {
+                                os.write(byteToWrite);
+                                IOUtils.fsync(tempDataPath, false);
+                            }
+                            Files.delete(tempDataPath);
+                            final long elapsedTime = currentTimeMillisSupplier.getAsLong() - executionStartTime;
+                            if (elapsedTime > slowPathLoggingThreshold.millis()) {
+                                logger.warn("health check of [{}] took [{}ms] which is above the warn threshold of [{}]",
+                                    path, elapsedTime, slowPathLoggingThreshold);
+                            }
                         }
-                        Files.delete(tempDataPath);
-                        final long elapsedTime = currentTimeMillisSupplier.getAsLong() - executionStartTime;
-                        if (elapsedTime > slowPathLoggingThreshold.millis()) {
-                            logger.warn("health check of [{}] took [{}ms] which is above the warn threshold of [{}]",
-                                path, elapsedTime, slowPathLoggingThreshold);
+                    } catch (Exception ex) {
+                        logger.error(new ParameterizedMessage("health check of [{}] failed", path), ex);
+                        if (currentUnhealthyPaths == null) {
+                            currentUnhealthyPaths = new HashSet<>(1);
                         }
+                        currentUnhealthyPaths.add(path);
                     }
-                } catch (Exception ex) {
-                    logger.error(new ParameterizedMessage("health check of [{}] failed", path), ex);
-                    if (currentUnhealthyPaths == null) {
-                        currentUnhealthyPaths = new HashSet<>(1);
-                    }
-                    currentUnhealthyPaths.add(path);
                 }
+            } catch (IllegalStateException e) {
```

Comment:
I'd prefer the `try{}` block only to contain the call to `nodeEnv.nodeDataPaths()`, would you reduce its scope? That way you don't need a local `lockAssertionFailed`, you can set `brokenLock` and exit immediately.